include/ginkgo/core/distributed/matrix.hpp

```diff
@@ -0,0 +1,544 @@
+                         .get()),
+                 static_cast<const LinOp*>(
+                     non_local_matrix_template
+                         .template create<ValueType, LocalIndexType>(exec)
+                         .get()))
+    {}
+
+    /**
+     * Creates an empty distributed matrix with specified type
+     * for local matricies.
+     *
+     * @note It internally clones the passed in matrix_template. Therefore, the
+     *       LinOp should be empty.
+     *
+     * @param exec  Executor associated with this matrix.
+     * @param comm  Communicator associated with this matrix.
+     * @param matrix_template  the local matrices will be constructed with the
+     *                         same runtime type.
+     */
+    explicit Matrix(std::shared_ptr<const Executor> exec,
+                    mpi::communicator comm, const LinOp* matrix_template);
+
+    /**
+     * Creates an empty distributed matrix with specified types for the local
+     * matrix and the non-local matrix.
+     *
+     * @note It internally clones the passed in local_matrix_template and
+     *       non_local_matrix_template. Therefore, those LinOps should be empty.
+     *
+     * @param exec  Executor associated with this matrix.
+     * @param comm  Communicator associated with this matrix.
+     * @param local_matrix_template  the local matrix will be constructed
+     *                               with the same runtime type.
+     * @param non_local_matrix_template  the non-local matrix will be
+     *                                   constructed with the same runtime type.
+     */
+    explicit Matrix(std::shared_ptr<const Executor> exec,
+                    mpi::communicator comm, const LinOp* local_matrix_template,
+                    const LinOp* non_local_matrix_template);
+
+    /**
+     * Starts a non-blocking communication of the values of b that are shared
+     * with other processors.
+     * @param local_b  The full local vector to be communicated. The subset of
+     *                 shared values is automatically extracted.
+     * @return  MPI request for the non-blocking communication.
+     */
+    mpi::request communicate(const local_vector_type* local_b) const;
+
+    void apply_impl(const LinOp* b, LinOp* x) const override;
+
+    void apply_impl(const LinOp* alpha, const LinOp* b, const LinOp* beta,
+                    LinOp* x) const override;
+
+private:
+    std::vector<comm_index_type> send_offsets_;
+    std::vector<comm_index_type> send_sizes_;
+    std::vector<comm_index_type> recv_offsets_;
+    std::vector<comm_index_type> recv_sizes_;
+    array<local_index_type> gather_idxs_;
+    array<global_index_type> non_local_to_global_;
+    gko::detail::DenseCache<value_type> one_scalar_;
+    gko::detail::DenseCache<value_type> host_send_buffer_;
+    gko::detail::DenseCache<value_type> host_recv_buffer_;
+    gko::detail::DenseCache<value_type> send_buffer_;
+    gko::detail::DenseCache<value_type> recv_buffer_;
```

Comment:
Can we maybe add these into structs to simplify readability and that also maybe reduces the number of parameters we need to pass through between functions.